lib/src/mt_rtcp.c

```diff
@@ -289,25 +254,35 @@ int mt_rtcp_rx_send_nack_packet(struct mt_rtcp_rx* rx) {
   rte_memcpy(rtcp->name, "IMTL", 4);
   uint16_t num_fci = 0;
 
-  struct mt_rtcp_nack_item *nack, *tmp_nack;
   struct mt_rtcp_fci* fci = rtcp->fci;
-  for (nack = MT_TAILQ_FIRST(&rx->nack_list); nack != NULL; nack = tmp_nack) {
-    tmp_nack = MT_TAILQ_NEXT(nack, next);
-    if (now > nack->expire_time) {
-      MT_TAILQ_REMOVE(&rx->nack_list, nack, next);
-      mt_rte_free(nack);
-      rx->stat_nack_drop_expire++;
-      continue;
+
+  /* check missing pkts with bitmap, fill fci fields */
+  uint16_t seq = rx->last_cont + 1;
+  uint16_t start = seq;
+  uint16_t miss = 0;
+  while (rtp_seq_num_cmp(seq, rx->last_seq - rx->seq_skip_window) < 0) {
+    if (!mt_bitmap_test(rx->seq_bitmap, seq % rx->seq_window_size)) {
+      miss++;
+    } else if (miss != 0) {
+      fci->start = htons(start);
+      fci->follow = htons(miss - 1);
+      fci++;
+      num_fci++;
+      rx->stat_rtp_lost_detected += miss;
+      start = seq + 1;
+      miss = 0;
     }
-    if (nack->retry_count == 0) continue;
-    fci->start = htons(nack->seq_id);
-    fci->follow = htons(nack->bulk);
+    seq++;
+  }
+  if (miss != 0) {
+    fci->start = htons(start);
+    fci->follow = htons(miss - 1);
     num_fci++;
-    fci++;
-    nack->retry_count--;
+    rx->stat_rtp_lost_detected += miss;
   }
-  if (num_fci > 32) {
-    dbg("%s(%s), too many nack items\n", __func__, rx->name);
+
+  if (num_fci == 0 || num_fci > 32) {
+    dbg("%s(%s), no / too many nack items %u\n", __func__, rx->name, num_fci);
```

Comment:
Cat we move the `num_fci` check before alloc mbuf? Sometimes in bad network case, it may always in such sequence `malloc, count num_fci, check num_fci fail, free mbuf`

Reply:
> Cat we move the `num_fci` check before alloc mbuf? Sometimes in bad network case, it may always in such sequence `malloc, count num_fci, check num_fci fail, free mbuf`

OK, then we need to fill the fci structures first and if num_fci within range copy them to mbuf.